text/XXX-unwind-FFI.md

```diff
@@ -0,0 +1,286 @@
+
+For now, the only valid `runtime` keys will be:
+
+* `default` - identical to `panic = 'unwind'` with no `runtime` selected.
+  This will use Rust's default unwinding runtime,
+  unless another crate in the final product's dependency graph
+specifies `panic = 'abort'`,
+  in which case the `abort` strategy will take precedence as usual.
+  which is not guaranteed to be compatible with native exceptions.
+* `native` - as discussed above, this will preserve the behavior of the current
+  implementation of `libpanic_unwind`.
+* `crate` - indicates that a non-`std` crate will use `#![panic_runtime]` to
+  provide the runtime.
+* `self` - indicates that this crate itself provides `#![panic_runtime]`.
+
+If the `native` or `crate` key is specified
+anywhere in a final product's dependency graph,
+no crate in that dependency graph may specify the `panic = abort` strategy;
+this mismatch will cause the build to fail.
+
+The function annotation `#[unwind(allowed)]` will only be permitted in crates
+that specify a non-default `panic.runtime`.
+
+Crates that do not use the `profile.foo.panic` option at all
+will remain compatible with any `profile.foo.panic` configuration
+used to generate the final product.
+
+For non-Cargo users, equivalent `rustc` flags will be provided
+(which will be how Cargo itself implements the option).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Unwinding for functions with the `#[unwind(allowed)]` annotation
+is performed as if the function were not marked `extern`.
+This annotation is not permitted on functions not marked `extern`.
+It has no observable effect unless the marked function `panic`s
+(e.g. it has no observable effect
+when a function returns normally or enters an infinite loop).
+The LLVM IR for such functions must not be marked `nounwind`.
+
+The compiler will have a new stable flag, `-C panic.runtime`,
+which will be required to enable the `#[unwind(allowed)]` annotation;
+as explained above, the flag will specify the expected source
+of the panic runtime to be used in the final (binary) product.
+If the source is `default`,
+the `libpanic_unwind` crate will provide the runtime,
+and the `#[unwind(allowed)]` annotation will not be permitted in that crate.
+
+Different values of `panic.runtime` will not be permitted for different crates
+in a single dependency graph,
+with the exception of `self`, which may only be used once
+in a dependency graph,
+and is only compatible with crates using the `crate` value.
+Because the `panic` runtime is lazily selected and linked,
+crates that do not specify a value for `panic.strategy`
+are compatible with crates using any of the four values.
+Crates that explicitly specify the `default` runtime, however,
+are not compatible with crates using other runtimes.
+
+In order to verify that all crates in a dependency graph
+are compatible in this way,
+new metadata will need to be added to the `rlib` format.
```

Comment:
This, like the concept of metadata, is a compiler implementation detail. Rather it is sufficient for a normative document to only specify what is checked, what is allowed and what is the expected outcomes in all sorts of situations are – just like you did in the section above.

Holistically following the RFC process (which it isn’t, but that is not the point), this section may force our hand in implementing something in a way that’s not necessarily the easiest/best/etc approach.

I recommend removing this paragraph entirely.

Reply:
Okay; I was trying to follow the example set by RFC 1513, which states,

> This codegen option will default to `unwind` if not specified (what happens
> today), and the value will be encoded into the crate metadata. This option is
> planned with extensibility in mind to future panic strategies if we ever
> implement some (return-based unwinding is at least one other possible option).

Would rephrasing to refer to "crate metadata" instead of "the `rlib` format" specifically be preferable to removing it?